File: Week_07-字典树-并差集-AVL树/212.单词搜索-ii.cpp
/*
 * @lc app=leetcode.cn id=212 lang=cpp
 *
 * [212] 单词搜索 II
 */

// @lc code=start
//有了这个Trie，看怎么利用，实现解题
class Trie {
private:
    bool isEnd;
    Trie* next[26] = {nullptr};                        //数组指针，表示26个指针元素组成的一维数组
public:
    /** Initialize your data structure here. */
    Trie() {                              //构造结点：1，初始化结束符isEnd，未结束；2，为新结点分配内存；
        isEnd = false;
        memset(next, 0, sizeof(next));
    }

    /** Inserts a word into the trie. */
    //从word的起始字符开始和trie中的next链接进行比对；
    //链接存在。沿着链接移动到树的下一个子层。算法继续搜索下一个键字符。
    //链接不存在。创建一个新的节点，并将它与父节点的链接相连，该链接与当前的键字符相匹
    void insert(string word) {
        Trie* node = this;            //指向当前结点
        for (char c : word) {
            if (node->next[c-'a'] == NULL) {   //如果指向空，那就构造一个新结点；
                node->next[c-'a'] = new Trie();
            }
            node = node->next[c-'a']; //指向下一个结点
        }
        node->isEnd = true;           //标记单词结束
    }
    
    /** Returns if the word is in the trie. */
    //在trie中，且是trie中的一个完整单词，而不仅是一个前缀；
    bool search(string word) {
        Trie* node = this;
        for (char c : word) {
            node = node->next[c-'a'];
            if (node == NULL) {
                return false;
            }
        }
        return node->isEnd;  //返回是否单词结束
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix) {
        Trie* node = this;
        for (char c : prefix) {
            node = node->next[c-'a'];
            if (node == NULL) {
                return false;   //返回没有找到完整的前缀
            }
        }
        return true;    //返回找到了完整的前缀
    }
};

class Solution {
public:
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        //先把words，写到trie中
        Trie trie;
        for (auto& w : words) {
            trie.insert(w);
        }
        vector<string> res;
        
        
        for (int row=0; row<board.size(); row++) {
            for (int col; col<board[0].size(); col++) {
                string line;
                backtrace(board, res, trie, line, row, col);
            }
        }
        return res;
    }
private:
    void backtrace(vector<vector<char>>& board, vector<string>& res, Trie& trie, string line, int row, int col) {
        if (trie.search(line)) {
            res.push_back(line);
            return;
        } else if (!trie.startsWith(line)) {
            return; 
        }
        char tmp = board[row][col];
        board[row][col] = 0;
        if(row>0) {
            backtrace(board, res, trie, line+(board[row-1][col]), row-1, col);
        }
        if (col>0) {
            backtrace(board, res, trie, line+(board[row][col-1]), row, col-1);
        }
        if (row<board.size()-1) {
            backtrace(board, res, trie, line+(board[row+1][col]), row+1, col);
        }
        if (col<board[0].size()-1) {
            backtrace(board, res, trie, line+(board[row][col+1]), row, col+1);
        }
        board[row][col] = tmp;
    }

};
// @lc code=end

File: Week_07-字典树-并差集-AVL树/212.单词搜索-ii.py
#
# @lc app=leetcode.cn id=212 lang=python
#
# [212] 单词搜索 II
#

# @lc code=start
from collections import defaultdict
    
class Solution(object):
    '''
    # 此方法超时!!!
    directs = [(1, 0), (-1, 0), (0, 1), (0,-1)]
    def findWords(self, board, words):
        """
        :type board: List[List[str]]
        :type words: List[str]
        :rtype: List[str]
        """
        ans = list()
        m = len(board)
        if m == 0: return ans
        n = len(board[0])
        for word in words:
            mark = [[0 for _ in range(n)] for _ in range(m)]
            inboard = False
            for i in range(m):
                for j in range(n):
                    if board[i][j] == word[0]:
                        #把该元素记为已使用
                        mark[i][j] = 1
                        if self.backtrack(i, j, mark, board, word[1:]) == True:
                            inboard = True
                        else:#回溯
                            mark[i][j] = 0
            if inboard: ans.append(word)
        return ans

    def backtrack(self, i, j, mark, board, word):
        if len(word) == 0:
            return True
        
        for direct in self.directs:
            cur_i = i + direct[0]
            cur_j = j + direct[1]
            if cur_i >= 0 and cur_i < len(board) and cur_j >= 0 and cur_j < len(board[0]) and board[cur_i][cur_j] == word[0]:
                # 如果是已经使用过的元素，忽略
                if mark[cur_i][cur_j] == 1:
                    continue
                # 将该元素标记为已使用
                mark[cur_i][cur_j] = 1
                if self.backtrack(cur_i, cur_j, mark, board, word[1:]) == True:
                    return True
                else:
                    # 回溯
                    mark[cur_i][cur_j] = 0
        return False
    '''

    # 针对上面dfs回溯超时：
    # 采用统计单词字母个数加速，比如某个单词要10个'e'，棋盘上一共9个'e'，那必然不能构成这个单词
    directs = [(1, 0), (-1, 0), (0, 1), (0,-1)]
    def findWords(self, board, words):
        """
        :type board: List[List[str]]
        :type words: List[str]
        :rtype: List[str]
        """
        ans = list()
        m = len(board)
        if m == 0: return ans
        n = len(board[0])
        #统计board中各元素数目
        dic_board = defaultdict(int)
        for i in range(m):
            for j in range(n):
                dic_board[board[i][j]] += 1
        
        for word in words:
            mark = [[0 for _ in range(n)] for _ in range(m)]
            dic_word = defaultdict(int)
            #统计word中各元素数目，并判断是否大于board中相应元素
            exceedboard = False
            for s in word:
                dic_word[s] += 1
            for k, v in dic_word.items():
                if v > dic_board[k]: 
                    exceedboard = True
                    break
            if exceedboard:
                continue

            inboard = False
            for i in range(m):
                for j in range(n):
                    if board[i][j] == word[0]:
                        #把该元素记为已使用
                        mark[i][j] = 1
                        if self.backtrack(i, j, mark, board, word[1:]) == True:
                            inboard = True
                        else:#回溯
                            mark[i][j] = 0
            if inboard: ans.append(word)
        return ans
    def backtrack(self, i, j, mark, board, word):
        if len(word) == 0:
            return True
        for direct in self.directs:
            cur_i = i + direct[0]
            cur_j = j + direct[1]
            if cur_i >= 0 and cur_i < len(board) and cur_j >= 0 and cur_j < len(board[0]) and board[cur_i][cur_j] == word[0]:
                # 如果是已经使用过的元素，忽略
                if mark[cur_i][cur_j] == 1:
                    continue
                # 将该元素标记为已使用
                mark[cur_i][cur_j] = 1
                if self.backtrack(cur_i, cur_j, mark, board, word[1:]) == True:
                    return True
                else:
                    # 回溯
                    mark[cur_i][cur_j] = 0
        return False

# @lc code=end

